lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1367 @@
+      attrStarts->add_ints(starts[b]);
+      attrEnds->add_ints(outs[b] + starts[b]);
+    }
+  }
+  return writeAllWithNode("Slice", node, proto);
+}
+
+llvm::Error ONNXModelWriter::writePow(const Node *node, GraphType &graph,
+                                      ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  proto->set_name(node->getName());
+  proto->add_input(node->getNthInput(PowNode::LHSIdx).getNode()->getName());
+  outputsToProto(node, proto);
+
+  // Find exponent from splat node
+  const auto *RHSN = node->getNthInput(PowNode::RHSIdx).getNode();
+  switch (RHSN->getKind()) {
+  case Kinded::Kind::SplatNodeKind: {
+    const auto *SN = llvm::dyn_cast<SplatNode>(RHSN);
+    RETURN_ERR_IF_NOT(SN, "Can't find Splat Node as part of Pow Node.");
+    float value = SN->getValue();
+    if (value == 0.5f) {
+      proto->set_op_type("Sqrt");
+    } else if (value == -1.0f) {
+      proto->set_op_type("Reciprocal");
+    } else {
+      RETURN_ERR_IF_NOT(false, "Splat Node Value is invalid.");
+    }
+    break;
+  }
+  default:
+    proto->add_input(RHSN->getName());
+    break;
+  }
+
+  reporter(RHSN);
+  return llvm::Error::success();
+}
+
+llvm::Error ONNXModelWriter::writeTopK(const Node *node, GraphType &graph,
+                                       ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *TKN = llvm::dyn_cast<TopKNode>(node);
+  RETURN_ERR_IF_NOT(TKN, "Invalid node type, expected TopKNode.");
+
+  auto *attr = proto->add_attribute();
+  attr->set_name("k");
+  attr->set_type(AttrType::INT);
+  attr->set_i(TKN->getK());
+
+  return writeAllWithNode("topK", node, proto);
+}
+
+llvm::Error ONNXModelWriter::writePRelu(const Node *node, GraphType &graph,
+                                        ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  proto->set_name(node->getName());
+  proto->set_op_type("PRelu");
+  proto->add_input(node->getNthInput(PReluNode::InputIdx).getNode()->getName());
+
+  const auto *slope = unwindBroadcastInput(
+      node->getNthInput(PReluNode::SlopeIdx).getNode(), reporter);
+  switch (slope->getKind()) {
+  case Kinded::Kind::SplatNodeKind: {
+    // Conversion a scalar to a tensor is required
+    Tensor scalar = {llvm::dyn_cast<SplatNode>(slope)->getValue()};
+    auto *tensorProto = graph.add_initializer();
+    tensorProto->set_name(slope->getName());
+    writeTensor(scalar, tensorProto);
+    proto->add_input(slope->getName());
+    reporter(slope);
+    break;
+  }
+  case Kinded::Kind::ReshapeNodeKind:
+    proto->add_input(
+        slope->getNthInput(ReshapeNode::InputIdx).getNode()->getName());
+    reporter(slope);
+    break;
+  default:
+    RETURN_ERR_IF_NOT(false,
+                      "Can't find Splat/Reshape Node as part of PRelu Node.");
+  }
+
+  outputsToProto(node, proto);
+  return llvm::Error::success();
+}
+
+llvm::Error ONNXModelWriter::writeGather(const Node *node, GraphType &graph,
```

Comment:
Many of these serialization methods seems to have a lot of boilerplate code. 

I wonder if it is possible to automate their serialization by means of auto-generating this code using NodeGen or something like it.

Reply:
Either  we need a new Glow loader then Glow exporter can be very much formalized and auto-generated, however if we want to use ONNX Loader (with some extension to support Glow specific nodes), it's not trivial to automate all special cases, for instance Conv requires Transposition of input and weights, whereas TransposeFC doesn't, etc.